### PR TITLE
Make the `Regex` type Sendable

### PR DESCRIPTION
**Pull request by natecook1000**
This is mostly trivial, with the exception of the lazy compiled regex. For that, we can use atomics to make storing the compiled program safe, even if it might result in an extra, thrown-away compilation.

- ~So that we don't have to import `swift-atomics`, I've subset out the `UnsafeAtomicLazyReference` type and its dependencies.~
- I've added a test that consistently traps due to concurrent mutation under `main`, but succeeds with these changes.
- This is now using the existing `_stdlib_atomicInitializeARCRef` function to synchronize initialization of the lazily compiled regex.

### CODE REVIEW

**Comment by natecook1000** [comment]
@swift-ci Please test

**Comment by natecook1000** [comment]
@swift-ci Please test

File: Sources/_StringProcessing/Regex/Core.swift [hunk 1]
```diff
@@ -68,20 +68,35 @@ extension Regex {
...
-  internal class Program {
+  internal final class Program: @unchecked Sendable {
     /// The underlying IR.
     ///
     /// FIXME: If Regex is the unit of composition, then it should be a Node instead,
     /// and we should have a separate type that handled both global options and,
     /// likely, compilation/caching.
     let tree: DSLTree
 
+    private final class ProgramBox {
```
Comment (milseman) [inline Sources/_StringProcessing/Regex/Core.swift:84]: Why do we need two boxes?

File: Package.swift [hunk 2]
```diff
@@ -60,9 +60,19 @@ let package = Package(
         .target(
             name: "_CUnicode",
             dependencies: []),
+        .target(
+            name: "_LazyAtomicShims",
+            dependencies: []),
+        .target(
+            name: "_LazyAtomic",
+            dependencies: ["_LazyAtomicShims"]),
```
[the remark this answers is not in the document]
Comment (natecook1000) [reply Package.swift:68]: > Why is this a separate target?
Good point, I can just move the Swift API into _StringProcessing.
> How does this affect compiler integration?
Not sure, how does it?

File: Sources/_StringProcessing/Regex/Core.swift (hunk 1 above)
Comment (natecook1000) [reply Sources/_StringProcessing/Regex/Core.swift:84]: `UnsafeAtomicLazyReference` is generic over an `AnyObject`. I guess we could skip that type and just store this as a `Optional<UnsafeMutablePointer<MEProgram<String>>>` — that would save one level of indirection, right?

File: Package.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (rxwei) [reply Package.swift:68]: _StringProcessing is built using make in the compiler repo with sources in the _StringProcessing directory. These new dependencies will break the build. Can these be added to _core_ as SPI instead, similar to the CUnicode SPI?

[the remark this answers is not in the document]
Comment (natecook1000) [reply Package.swift:68]: That should be fine — I see where those sources are getting added 👍🏻

**Review by rxwei — APPROVED** [review]

**Comment by natecook1000** [comment]
@swift-ci Please test

File: Sources/_StringProcessing/Regex/Core.swift (hunk 1 above)
Comment (natecook1000) [reply Sources/_StringProcessing/Regex/Core.swift:83]: Updated implementation only uses one layer of indirection 👍🏻

**Pull request merged**